File: packages/igloo-ts/src/widgets/scrollable.ts
import { Scrollable as WitScrollable } from 'iced:app/scrollable@0.1.0';
import type { Direction, Scrollbar, Anchor } from 'iced:app/scrollable@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Pixels } from 'iced:app/shared@0.1.0';
import type { Viewport } from 'iced:app/message-types@0.1.0';
import { Element, toElement, type ElementLike, type IntoElement } from '../element.js';
import { pushViewport } from '../callbacks.js';

export type { Direction, Scrollbar } from 'iced:app/scrollable@0.1.0';

/**
 * Builder for creating Scrollable widgets.
 * A Scrollable wraps content that can be scrolled.
 *
 * @example
 * ```typescript
 * const scrollable = Scrollable.new(
 *   Column.new().push(Text.new('Item 1')).push(Text.new('Item 2')),
 * ).height(Length.fixed(300));
 * ```
 *
 * @typeParam Msg - The message type of the content plus `onScroll`, inferred from `new`.
 */
export class Scrollable<Msg = never> implements IntoElement<Msg> {
  private raw: WitScrollable;

  private constructor(content: ElementLike<Msg>) {
    this.raw = new WitScrollable(toElement(content).inner);
  }

  /** Create a new Scrollable builder with the given content */
  static new<const Msg = never>(content: ElementLike<Msg>): Scrollable<Msg> {
    return new Scrollable(content);
  }

  /** Set the width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Set the height */
  height(height: Length): this {
    this.raw.height(height);
    return this;
  }

  /** Set the message to emit when scrolling occurs, given the new viewport */
  onScroll<const M>(mapper: (viewport: Viewport) => M): Scrollable<Msg | M> {
    this.raw.onScroll(pushViewport(mapper));
    return this as unknown as Scrollable<Msg | M>;
  }

  /** Set the scroll direction and scrollbar configuration */
  direction(direction: Direction): this {
    this.raw.direction(direction);
    return this;
  }

  /** Configure for vertical scrolling only */
  vertical(scrollbar: Scrollbar = {}): this {
    this.raw.direction({ tag: 'vertical', val: scrollbar });
    return this;
  }

  /** Configure for horizontal scrolling only */
  horizontal(scrollbar: Scrollbar = {}): this {
    this.raw.direction({ tag: 'horizontal', val: scrollbar });
    return this;
  }

  /** Configure for both vertical and horizontal scrolling */
  both(verticalScrollbar: Scrollbar = {}, horizontalScrollbar: Scrollbar = {}): this {
    this.raw.direction({ tag: 'both', val: [verticalScrollbar, horizontalScrollbar] });
    return this;
  }

  /** Convert to Element */
  intoElement(): Element<Msg> {
    return new Element(WitScrollable.intoElement(this.raw));
  }
}

/**
 * Helper to create scrollbar configuration.
 */
export const ScrollbarConfig = {
  /** Create a default scrollbar config */
  default(): Scrollbar {
    return {};
  },

  /** Create a scrollbar config with custom settings */
  custom(options: {
    width?: Pixels;
    margin?: Pixels;
    scrollerWidth?: Pixels;
    anchor?: Anchor;
    spacing?: Pixels;
  }): Scrollbar {
    return options;
  },
} as const;

File: packages/igloo-py/src/igloo_py/widgets/scrollable.py
"""Scrollable widget builder."""

from collections.abc import Callable
from typing import TYPE_CHECKING, Any, TypeVar

from wit_world.imports.message_types import Viewport
from wit_world.imports.scrollable import (
    Direction_Both,
    Direction_Horizontal,
    Direction_Vertical,
    Scrollbar,
)
from wit_world.imports.scrollable import (
    Scrollable as WitScrollable,
)

from ..callbacks import push_viewport
from ..element import Element, ElementLike, IntoElement, to_element

if TYPE_CHECKING:
    from ..types.length import WitLength

Msg = TypeVar("Msg")


class Scrollable(IntoElement):
    """
    Builder for creating Scrollable widgets.
    A Scrollable wraps content that can be scrolled.

    Example:
        scrollable = Scrollable.new(
            Column.new()
                .push(Text.new("Item 1"))
                .push(Text.new("Item 2"))
                # ... more items
        ).height(Length.fixed(300))
    """

    def __init__(self, content: ElementLike) -> None:
        self._raw = WitScrollable(to_element(content).inner)

    @classmethod
    def new(cls, content: ElementLike) -> "Scrollable":
        """Create a new Scrollable builder with the given content."""
        return cls(content)

    def width(self, width: "WitLength") -> "Scrollable":
        """Set the width."""
        self._raw.width(width)
        return self

    def height(self, height: "WitLength") -> "Scrollable":
        """Set the height."""
        self._raw.height(height)
        return self

    def on_scroll(self, mapper: Callable[[Viewport], Msg]) -> "Scrollable":
        """
        Set the message to emit when scrolling occurs.
        The Message will have tag 'viewport' with scroll position info.
        """
        self._raw.on_scroll(push_viewport(mapper))
        return self

    def direction(self, direction: Any) -> "Scrollable":
        """Set the scroll direction and scrollbar configuration."""
        self._raw.direction(direction)
        return self

    def vertical(self, scrollbar: dict[str, Any] | None = None) -> "Scrollable":
        """Configure for vertical scrolling only."""
        sb = Scrollbar(
            width=scrollbar.get("width") if scrollbar else None,
            margin=scrollbar.get("margin") if scrollbar else None,
            scroller_width=scrollbar.get("scroller_width") if scrollbar else None,
            anchor=scrollbar.get("anchor") if scrollbar else None,
            spacing=scrollbar.get("spacing") if scrollbar else None,
        )
        self._raw.direction(Direction_Vertical(sb))
        return self

    def horizontal(self, scrollbar: dict[str, Any] | None = None) -> "Scrollable":
        """Configure for horizontal scrolling only."""
        sb = Scrollbar(
            width=scrollbar.get("width") if scrollbar else None,
            margin=scrollbar.get("margin") if scrollbar else None,
            scroller_width=scrollbar.get("scroller_width") if scrollbar else None,
            anchor=scrollbar.get("anchor") if scrollbar else None,
            spacing=scrollbar.get("spacing") if scrollbar else None,
        )
        self._raw.direction(Direction_Horizontal(sb))
        return self

    def both(
        self,
        vertical_scrollbar: dict[str, Any] | None = None,
        horizontal_scrollbar: dict[str, Any] | None = None,
    ) -> "Scrollable":
        """Configure for both vertical and horizontal scrolling."""
        vsb = Scrollbar(
            width=vertical_scrollbar.get("width") if vertical_scrollbar else None,
            margin=vertical_scrollbar.get("margin") if vertical_scrollbar else None,
            scroller_width=vertical_scrollbar.get("scroller_width") if vertical_scrollbar else None,
            anchor=vertical_scrollbar.get("anchor") if vertical_scrollbar else None,
            spacing=vertical_scrollbar.get("spacing") if vertical_scrollbar else None,
        )
        hsb = Scrollbar(
            width=horizontal_scrollbar.get("width") if horizontal_scrollbar else None,
            margin=horizontal_scrollbar.get("margin") if horizontal_scrollbar else None,
            scroller_width=horizontal_scrollbar.get("scroller_width")
            if horizontal_scrollbar
            else None,
            anchor=horizontal_scrollbar.get("anchor") if horizontal_scrollbar else None,
            spacing=horizontal_scrollbar.get("spacing") if horizontal_scrollbar else None,
        )
        self._raw.direction(Direction_Both((vsb, hsb)))
        return self

    def into_element(self) -> Element:
        """Convert to Element."""
        return Element(WitScrollable.into_element(self._raw))
